Allow erc20 token transfers

AI-Agent-D opened the pull request.

Kudo-Dev-tech commented on src/actions/transfer.ts:
```diff
@@ -7,29 +7,75 @@ import {
+import { transferTemplate } from "../templates";
+import type { Transaction, TransferParams } from "../types";
+
+export const ABIEncoding = async(amount: string, recipientAddress: string, tokenDecimals: string) => {
```
make sure to format this

Kudo-Dev-tech replied:
Also rename this function to `getTransferData`.  Generally prefer to start function names with a verb.

Kudo-Dev-tech replied:
Also specify the return type.  I'll let you try figure out what it should be

Kudo-Dev-tech commented on src/actions/transfer.ts:
```diff
@@ -7,29 +7,75 @@ import {
+        "constant": false,
+        "inputs": [
+            {
+                "name": "_sireId",
```
```suggestion
                "name": "to",
```

Kudo-Dev-tech commented on src/actions/transfer.ts:
```diff
@@ -7,29 +7,75 @@ import {
+                "type": "address"
+            },
+            {
+                "name": "_matronId",
```
```suggestion
                "name": "amount",
```

Kudo-Dev-tech commented on src/actions/transfer.ts:
```diff
@@ -7,29 +7,75 @@ import {
+
+  let data = ''
+  
+  if (amount === '0') {
```
Do a check at the beginning of the function to terminate it early if `amount > 0`

```
if (amount > 0) return "0x"
```

You can then

1.  Get rid of `let data = ''` above
2. Remove the `if-else`

Kudo-Dev-tech commented on src/actions/transfer.ts:
```diff
@@ -7,29 +7,75 @@ import {
+      params.data = "0x";
+    }
+
+    if ((params.amount === "0" && params.toAddress === params.recipientAddress) || (params.amount === "0" && params.tokenDecimals === "0")) {
```
When is `tokenDecimals` 0?

Kudo-Dev-tech replied:
Also this `if` statement can be refactored to be cleaner

```
if (params.amount === "0" && (params.toAddress === params.recipientAddress || params.tokenDecimals === "0"))
```

AI-Agent-D replied:
Token decimals equal 0 when there is a 0 transfer for ERC20 coins:

```
{
     amount: "0",
     tokenDecimals: "0",
     recipientAddress: "0xABC",
...

}
```

This is because the prompt calculates the tokenDecimals directly from the past messages and is instructed to put 0 for ERC20 tokens on the 'amount' field. Therefore, the only field that changes is the tokenDecimals in the ERC20 case.

Kudo-Dev-tech commented on src/actions/transfer.ts:
```diff
@@ -7,29 +7,75 @@ import {
+        recipientAddress,
+        tokenDecimals,
+      ]
+    }) as `0x${string}`
```
```suggestion
    }) as Hex
```

You may have to import `Hex`

Kudo-Dev-tech commented on src/actions/transfer.ts:
```diff
@@ -128,13 +182,20 @@ export const transferAction: Action = {
       const transferResp = await action.transfer(paramOptions);
       if (callback) {
+
+        if (paramOptions.amount)
```
Why do we need to do this check?

AI-Agent-D replied:
I think its because if the amount is an empty string: ```""``` then the transaction will fail (silently)

AI-Agent-D replied:
Removed

Kudo-Dev-tech commented on src/types/index.ts:
```diff
@@ -67,6 +69,9 @@ export interface TransferParams {
   toAddress: Address;
   amount: string;
```
We should update the type to be `bigint` here as we use it for numerical comparisons later in the code.

Kudo-Dev-tech commented on src/types/index.ts:
```diff
@@ -67,6 +69,9 @@ export interface TransferParams {
   data?: `0x${string}`;
+  tokenDecimals: string,
+  token: string,
+  recipientAddress: `0x${string}`;
```
```suggestion
  recipientAddress: Address;
```

Kudo-Dev-tech commented on src/actions/transfer.ts:
```diff
@@ -81,13 +128,20 @@ const buildTransferDetails = async (
   });
 
-  const parsedXml = parseKeyValueXml(xmlResponse);
+  let parsedXml = parseKeyValueXml(xmlResponse) as Omit<TransferParams, "data">;
```
Why change this to `let`?

Kudo-Dev-tech commented on src/actions/transfer.ts:
```diff
@@ -81,13 +128,20 @@ const buildTransferDetails = async (
   }
 
-  const transferDetails = parsedXml as unknown as TransferParams;
+  // Write th VIEM function here
```
Remove lingering comments

Kudo-Dev-tech commented on src/actions/transfer.ts:
```diff
@@ -128,13 +182,20 @@ export const transferAction: Action = {
+        if (paramOptions.amount)
         callback({
-          text: `Successfully transferred ${paramOptions.amount} tokens to ${paramOptions.toAddress}\nTransaction Hash: ${transferResp.hash}`,
+          text: `Successfully transferred ${paramOptions.amount} tokens to ${paramOptions.toAddress} Transaction Hash: ${transferResp.hash}`,
```
This won't work for ERC20 transactions.  It will say something like

`params.amount` will always be 0
`paramsOptions.toAddress` will always be the ERC20 token's address

Kudo-Dev-tech commented:
0.00001 will still evaluate to a big int if you convert it to 10e18

AI-Agent-D commented:
I changed the amount somewhat. I was struggling because of the case BigInt(0.0000001) evaluated to an error. Therefore, I converted the amount to the tokenDecimals * amount to make the BigInt.

Kudo-Dev-tech commented on src/actions/transfer.ts:
```diff
@@ -7,56 +7,100 @@ import {
+import { supportedChains } from "@lifi/data-types";
+import { printer } from "prettier/doc.js";
+
+export const getTransferData = async (amountDecimals: bigint, recipientAddress: string): Promise<Hex> => {
```
Name it `amount`

Kudo-Dev-tech replied:
```suggestion
export const getTransferData = async (amountDecimals: bigint, recipientAddress: Address): Promise<Hex> => {
```

Kudo-Dev-tech commented on src/actions/transfer.ts:
```diff
@@ -7,56 +7,100 @@ import {
 
 // Exported for tests
+
+const isNativeTransfer = (transferParams: TransferParams) => {
```
```suggestion
const isNativeTransfer = (transferParams: TransferParams) => transferParams.toAddress === transferParams.recipientAddress
```

Kudo-Dev-tech commented on src/actions/transfer.ts:
```diff
@@ -82,12 +126,25 @@ const buildTransferDetails = async (
+    token: parsedXml.token
+  } as TransferParams;
+
+  transferDetails.amount = BigInt(parseUnits(parsedXml.amount, transferDetails.tokenDecimals));
```
`parseUnits` already returns a `BigInt`
```suggestion
  transferDetails.amount = parseUnits(parsedXml.amount, transferDetails.tokenDecimals);
```

Kudo-Dev-tech commented on src/tests/transfer.test.ts:
```diff
@@ -57,186 +57,335 @@ describe('Transfer Action', () => {
-      it('should perform actual transfer if funded wallet is available', async () => {
+    it("should work with Base Sepolia testnet (for ERC20 - USDC)", async () => {
+      const balance = await wp.getWalletBalanceForChain("baseSepolia");
+      console.log(`Base Sepolia balance: ${balance} ETH`);
```
This should log out the USDC balance instead of ETH balance

Kudo-Dev-tech commented on src/tests/transfer.test.ts:
```diff
@@ -57,186 +57,335 @@ describe('Transfer Action', () => {
+
+      const usdcAddress = "0x1c7D4B196Cb0C7B01d743Fbc6116a902379C7238"
+
+      if (balance && parseFloat(balance) > 0.001) {
```
Make sure to double check this.  If `balance` is returned as 6dp then this will always return true.

AI-Agent-D replied:
It will always return the actual amount => Not a problem.

Kudo-Dev-tech commented on src/tests/transfer.test.ts:
```diff
@@ -57,186 +57,335 @@ describe('Transfer Action', () => {
+          recipientAddress: receiver.address,
+          token: "USDC",
+          tokenDecimals: 6,
+          data: await getTransferData(0, receiver.address, 6) as Hex
```
We are transferring 0 USDC?

AI-Agent-D replied:
Fixed

Kudo-Dev-tech commented on src/tests/transfer.test.ts:
```diff
@@ -57,186 +57,335 @@ describe('Transfer Action', () => {
+        const result = await ta.transfer({
+          fromChain: "baseSepolia" as any,
+          toAddress: usdcAddress,
+          amount: 1n,
```
Why is it `1n`?

Kudo-Dev-tech commented on src/tests/transfer.test.ts:
```diff
@@ -57,186 +57,335 @@ describe('Transfer Action', () => {
+      const balance = await wp.getWalletBalanceForChain("sepolia");
+      console.log(`Sepolia balance: ${balance} ETH`);
+
+      const usdcAddress = "0x1c7D4B196Cb0C7B01d743Fbc6116a902379C7238"
```
Assign this to a `constant` variable somewhere.  This is used 4 times in this test file.

Kudo-Dev-tech replied:
Also make sure it's the right address for the right chain.

Kudo-Dev-tech commented on src/tests/transfer.test.ts:
```diff
@@ -57,186 +57,335 @@ describe('Transfer Action', () => {
+            fromChain: "sepolia" as any,
             toAddress: receiver.address,
-            amount: '0.001', // 0.001 ETH
+            amount: parseEther("0.0001"), // 0.001 ETH
```
```suggestion
            amount: parseEther("0.0001"), // 0.0001 ETH
```

Kudo-Dev-tech commented on src/actions/transfer.ts:
```diff
@@ -82,12 +126,25 @@ const buildTransferDetails = async (
   });
 
   const parsedXml = parseKeyValueXml(xmlResponse);
```
Implement another function to validate the contents of `parsedXml` e.g `recipientAddress` is not the zero address

Kudo-Dev-tech commented on src/actions/transfer.ts:
```diff
@@ -82,12 +126,25 @@ const buildTransferDetails = async (
+  } as TransferParams;
+
+  transferDetails.amount = BigInt(parseUnits(parsedXml.amount, transferDetails.tokenDecimals));
+  transferDetails.data = await getTransferData(transferDetails.amount, transferDetails.recipientAddress);
```
```suggestion
  transferDetails.data = await getTransferCalldata(transferDetails.amount, transferDetails.recipientAddress);
```

Rename the function name

aiagentdev-1990 commented on src/providers/walletERC20.ts:
```diff
@@ -0,0 +1,117 @@
+import { initWalletProvider } from "./wallet";
+import { type SupportedChain } from "src/types";
+
+export const getWalletERC20Balance = async (
```
This should no longer be here

aiagentdev-1990 commented on src/providers/walletERC20.ts:
```diff
@@ -0,0 +1,117 @@
+  }
+};
+
+const TOKEN_ADDRS = {
```
We should remove this and get the token address and token decimals using an LLM

aiagentdev-1990 commented on src/providers/wallet.ts:
```diff
@@ -145,12 +150,92 @@ export class WalletProvider {
     }
   }
 
+  async getWalletERC20Balance(
```
Combine this into one function with [this](https://github.com/AI-Agent-D/plugin-evm/pull/1/files#diff-e3141cde9556c98e2ad0299d3cbd3919da62861f2e3996b78979d7516ccedf12R201).

aiagentdev-1990 commented on src/providers/walletERC20.ts:
```diff
@@ -0,0 +1,117 @@
+          const client = walletProvider.getPublicClient(
+            chain as unknown as SupportedChain,
+          );
+          const balance = await getWalletERC20Balance(
```
```suggestion
          const balance = await walletProvider.getWalletERC20Balance(
```

Remember that we moved the function into `walletProvider`.

AI-Agent-D commented:
I 'fixed' the problem by adding in a custom mock state by adding in a 'recentMessages' field.

aiagentdev-1990 commented on src/actions/transfer.ts:
```diff
@@ -77,6 +140,8 @@ const buildTransferDetails = async (
     template: transferTemplate,
   });
 
+  console.log(context);
```
Remove unused log

aiagentdev-1990 commented on src/actions/transfer.ts:
```diff
@@ -87,7 +152,22 @@ const buildTransferDetails = async (
+  } as TransferParams;
+
+  transferDetails.amount = parseUnits(parsedXml.amount, transferDetails.tokenDecimals);
+  transferDetails.data = await getTransferCallData(
```
This needs to be empty (`0x`) if it is a native transfer

aiagentdev-1990 commented on src/actions/transfer.ts:
```diff
@@ -87,7 +152,22 @@ const buildTransferDetails = async (
+    transferDetails.recipientAddress
+  );
+
+  checkTransferDetails(transferDetails);
```
It is good practice to error as early as possible.  In this case we should validate `parsedXml` instead of `transferDetails` and error before [this](https://github.com/copra-finance/plugin-evm/pull/1/files#diff-bfe7fa7cbe814c62316544db1a7111e18a755c457efd3b219995f0e181e602ecR155).

aiagentdev-1990 commented on src/actions/transfer.ts:
```diff
@@ -134,12 +214,12 @@ export const transferAction: Action = {
       if (callback) {
         callback({
-          text: `Successfully transferred ${paramOptions.amount} tokens to ${paramOptions.toAddress}\nTransaction Hash: ${transferResp.hash}`,
+          text: `Successfully transferred ${formatUnits(paramOptions.amount, paramOptions.tokenDecimals)} tokens to ${paramOptions.recipientAddress} Transaction Hash: ${transferResp.hash}`,
```
Is there a test that shows the `text` field as being correct?

AI-Agent-D replied:
Currently not. In fact the  transferAction does not have any tests.

aiagentdev-1990 commented on src/providers/walletERC20.ts:
```diff
@@ -0,0 +1,61 @@
+        Object.keys(tokenAddresses).map(async (tokenName) => {
+          const { tokenAddress, tokenDecimals } = tokenAddresses[tokenName];
+
+          console.log(`Querying balance for ${tokenName} at address ${tokenAddress}`);
```
Use `elizaLogger.log` instead of `console.log`

aiagentdev-1990 commented on src/providers/walletERC20.ts:
```diff
@@ -0,0 +1,104 @@
+import { type SupportedChain } from 'src/types';
+import { searchAddressTokenTemplate } from 'src/templates';
+
+interface TokenData {
```
Format this by running `bun format`

aiagentdev-1990 commented on src/providers/walletERC20.ts:
```diff
@@ -0,0 +1,104 @@
+  // Format in the token data field.
+
+  if (!state) {
+    state = (await runtime.composeState(_message)) as State;
```
```suggestion
    state = await runtime.composeState(_message)
```

No need to do `as State`.  `composeState` already returns an instance of type `State`.

aiagentdev-1990 commented on src/providers/walletERC20.ts:
```diff
@@ -0,0 +1,104 @@
+    try {
+      const walletProvider = await initWalletProvider(runtime);
+      const agentName = state?.agentName || 'The agent';
+      const tokenAddresses = await getTokenDecimalsAddress(runtime, _message, state)
```
Name this variable better i.e `tokenDataByChain`

aiagentdev-1990 commented on src/providers/walletERC20.ts:
```diff
@@ -0,0 +1,104 @@
+        queriedChainBalances += `${agentName}'s wallet holdings consists of ${balances} on chain ${chain}`
+    }
+
+      return {
```
Just return 

```
{
  text: queriedChainBalances
}
```

if `data` and `values` are empty

aiagentdev-1990 commented on src/providers/walletERC20.ts:
```diff
@@ -0,0 +1,104 @@
+      };
+
+    } catch (error) {
+      console.error('Error in EVM wallet provider:', error);
```
Use `elizaLogger`

aiagentdev-1990 commented on src/providers/walletERC20.ts:
```diff
@@ -0,0 +1,104 @@
+
+      for (const currentChain in tokenAddresses) {
+        const chain = currentChain as unknown as SupportedChain
+
```
Log the chain we are currently querying balances for

aiagentdev-1990 commented on src/providers/walletERC20.ts:
```diff
@@ -0,0 +1,104 @@
+      
+      let queriedChainBalances = ``
+
+      for (const currentChain in tokenAddresses) {
```
This will cause the code to run slower as it will do every loop sequentially.  The best way is to also use `Promise.all` so that it happens concurrently.

```
  const queriedChainBalances = await Promise.all(
      Object.entries(tokenDataByChain).map(async ([chainName, tokenDataByToken]) => {
        const tokenBalances = await Promise.all(
          Object.entries(tokenDataByToken).map(
            async ([tokenSymbol, { tokenAddress, tokenDecimals }]) => {
              elizaLogger.log(`Querying balance for ${tokenSymbol} at address ${tokenAddress}`);

              const balance = await walletProvider.getWalletBalanceForERC20(
                chainName as SupportedChain,
                tokenDecimals,
                tokenAddress
              );
              return `${balance} ${tokenSymbol}`;
            }
          )
        );

        return `${chainName} Balances: ${tokenBalances.join(', ')}`;
      })
    );
```

This will return a `string[]` that looks like something like

```
["BASE balances:  1 USDC, 0.5 USDC.e", "Arbitrum balances:  10 USDC, 2 ARB"]
```

You can then use `join` to combine the array into a string when you return it in the `text` field.

```
{
   text: `${agentName}'s EVM Wallet Address: ${walletProvider.account.address}\n ${queriedChainBalances.join('\n')}`,
}
```

`.join` will combine the elements into a single string and add a line break between each entry

aiagentdev-1990 commented on src/providers/walletERC20.ts:
```diff
@@ -0,0 +1,104 @@
+        })
+        );
+
+        queriedChainBalances += `${agentName}'s wallet holdings consists of ${balances} on chain ${chain}`
```
At this point `balances` is still an array so the formatting will be off as it will surround `balances` with `[]`.

Example
```
DegenSpartanAI's wallet holdings consists of [1 USDC, 2 ARB] on chain Arbitrum
```

aiagentdev-1990 commented on src/providers/walletERC20.ts:
```diff
@@ -0,0 +1,100 @@
+  _message: Memory,
+  state?: State
+): Promise<TokenData> => {
+  // Look through all recent messages and figure out which tokens have been mentioned and on which chains.
```
Remove unused comments

aiagentdev-1990 commented on vitest.config.ts:
```diff
@@ -0,0 +1,7 @@
+    test: {
+        exclude: ["**/tests/optionalTests/**"],
+    }
+})
```
Add extra line at the bottom.  It's good practice to do so.  [Article](https://sourabhkr.medium.com/why-should-text-files-end-with-a-newline-c2e08d4896b8).

aiagentdev-1990 commented on src/tests/wallet.test.ts:
```diff
@@ -258,4 +266,144 @@ describe('Wallet Provider', () => {
+      expect(result.data).toEqual(undefined)
+    })
+
+    it.only('should return the tokens on different test nets (baseSepolia and Sepolia)', async () => {
```
Remove `.only`

aiagentdev-1990 commented on .prettierignore:
```diff
@@ -0,0 +1,29 @@
+# Misc
+.DS_Store
+.env
+.env.*
```
Add trailing line

aiagentdev-1990 commented on .prettierrc:
```diff
@@ -0,0 +1,8 @@
+    "trailingComma": "es5",
+    "printWidth": 100,
+    "endOfLine": "lf"
+  }
```
Add trailing line

aiagentdev-1990 commented on src/tests/wallet.test.ts:
```diff
@@ -1,9 +1,17 @@
+import dotenv from 'dotenv';
+import { getToken } from '@lifi/sdk';
+
+dotenv.config();
```
This won't work in the pipeline as there isn't a `.env` file when it runs there.  Just follow what they've been doing [here](https://github.com/copra-finance/plugin-evm/pull/1/files#diff-e0ad27363c4cb4702a53ff2b10fe42eac0760bd499c698c5aeebaf01ff9237b3R17).

aiagentdev-1990 commented on src/tests/transfer.test.ts:
```diff
@@ -178,7 +317,11 @@ describe('Transfer Action', () => {
             fromChain: 'sepolia' as any,
             toAddress: receiver.address,
-            amount: '0.001', // 0.001 ETH
+            amount: parseEther('0.0001'), // 0.0001 ETH
```
```suggestion
            amount: parseEther('0.0001'),    
```

Remove comment as it's obvious it's `0.0001` ETH

aiagentdev-1990 commented on src/tests/transfer.test.ts:
```diff
@@ -149,13 +214,87 @@ describe('Transfer Action', () => {
+        const result = await ta.transfer({
+          fromChain: 'baseSepolia' as any,
+          toAddress: usdcAddressBase,
+          amount: 1000000n,
```
```suggestion
          amount: parseUnits(1, 6) // 1 USDC
```

Use `parseUnits` instead of hardcoding amounts

aiagentdev-1990 commented on src/tests/transfer.test.ts:
```diff
@@ -149,13 +214,87 @@ describe('Transfer Action', () => {
+
+        expect(result.hash).toMatch(/^0x[a-fA-F0-9]{64}$/);
+        expect(result.to).toBe(usdcAddressBase);
+        expect(result.data.startsWith('0x')).toBe(true);
```
Test that the encoded ABI is actually correct instead of just having it be `0x`.

aiagentdev-1990 commented on src/tests/transfer.test.ts:
```diff
@@ -86,32 +116,59 @@ describe('Transfer Action', () => {
     describe('Network-specific transfers', () => {
       it('should work with Sepolia testnet', async () => {
         const balance = await wp.getWalletBalanceForChain('sepolia');
+        console.log(balance);
```
Remove unused log
